File: services/gateway/app/models/cliente.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import Column, DateTime, Enum, Float, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.db.session import Base


class PerfilCliente(str, enum.Enum):
    FIEL = "fiel"
    ABANDONO = "abandono"
    ESQUECIDO = "esquecido"
    ECONOMICO = "economico"


class Cliente(Base):
    __tablename__ = "clientes"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    nome = Column(String(180), nullable=False)

    # CPF nunca em texto claro:
    #   - cpf_encrypted: Fernet (não-determinístico, garante confidencialidade)
    #   - cpf_hash: HMAC-SHA256(pepper, cpf) determinístico para lookup
    cpf_encrypted = Column(Text, nullable=False)
    cpf_hash = Column(String(64), unique=True, nullable=False, index=True)

    # Email e telefone também em Fernet (sem lookup determinístico).
    email_encrypted = Column(Text, nullable=True)
    telefone_encrypted = Column(Text, nullable=True)

    regiao = Column(String(40), nullable=False)
    perfil = Column(
        Enum(PerfilCliente, name="perfil_cliente", values_callable=lambda x: [e.value for e in x]),
        nullable=True,
        index=True,
    )
    score_risco = Column(Float, nullable=True)

    criado_em = Column(DateTime(timezone=True), default=datetime.utcnow, nullable=False, index=True)
    classificado_em = Column(DateTime(timezone=True), nullable=True)

    # LGPD: marcador de anonimização.
    anonimizado_em = Column(DateTime(timezone=True), nullable=True)

    veiculos = relationship("Veiculo", back_populates="cliente", cascade="all, delete-orphan")
    leads = relationship("Lead", back_populates="cliente", cascade="all, delete-orphan")

File: services/core/src/main/java/com/previopls/entity/Cliente.java
package com.previopls.entity;

import com.previopls.crypto.EncryptedStringConverter;
import com.previopls.entity.enums.PerfilCliente;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "clientes", indexes = {
        @Index(name = "ix_clientes_cpf", columnList = "cpf", unique = true),
        @Index(name = "ix_clientes_perfil", columnList = "perfil")
})
@Getter
@Setter
@NoArgsConstructor
public class Cliente {

    @Id
    @GeneratedValue
    @Column(columnDefinition = "uuid")
    private UUID id;

    @Column(nullable = false, length = 180)
    private String nome;

    @Column(nullable = false, unique = true, length = 14)
    private String cpf;

    @Convert(converter = EncryptedStringConverter.class)
    @Column(columnDefinition = "text")
    private String email;

    @Convert(converter = EncryptedStringConverter.class)
    @Column(columnDefinition = "text")
    private String telefone;

    @Column(nullable = false, length = 40)
    private String regiao;

    @Enumerated(EnumType.STRING)
    @Column(length = 20)
    private PerfilCliente perfil;

    @Column(name = "score_risco")
    private Double scoreRisco;

    @Column(name = "criado_em", nullable = false, updatable = false)
    private Instant criadoEm = Instant.now();

    @Column(name = "classificado_em")
    private Instant classificadoEm;

    @Column(name = "retencao_ate")
    private Instant retencaoAte;

    @OneToMany(mappedBy = "cliente", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<Veiculo> veiculos = new ArrayList<>();

    @OneToMany(mappedBy = "cliente", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<Lead> leads = new ArrayList<>();
}
